README.md
# Project Woothee

Project Woothee is multi-language user-agent strings parsers.

## Why new project?

We needs just same logic over 2 or more programming languages, for use on various frameworks, middlewares and environments.

Most important data of this project is only single set of return values, and set of test cases, for equality of results of another languages implementations.

Implemantations:

  * Java (and Hive UDF)
  * Perl
  * Ruby
  * Python
  * Javascript (Node.js or browser)

## SYNOPSIS
in Java: (use java/woothee.jar)

```java
// import is.tagomor.woothee.Classifier;
// import is.tagomor.woothee.DataSet;
Map r = Classifier.parse("user agent string");
    
r.get("name")
// => name of browser (or string like name of user-agent

r.get("category")
// => "pc", "smartphone", "mobilephone", "appliance", "crawler", "misc", "unknown"

r.get("os")
// => os from user-agent, or carrier name of mobile phones

r.get("version");
// => version of browser, or terminal type name of mobile phones
```

in Perl: (cpanm Woothee)

```perl
use Woothee;
Woothee::parse("Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 6.1; Trident/4.0)");
# => {'name'=>"Internet Explorer", 'category'=>"pc", 'os'=>"Windows 7", 'version'=>"8.0", 'vendor'=>"Microsoft"}
```

in Ruby: (gem install woothee)

```ruby
require 'woothee'
Woothee.parse("Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 6.1; Trident/4.0)")
# => {:name=>"Internet Explorer", :category=>:pc, :os=>"Windows 7", :version=>"8.0", :vendor=>"Microsoft"}
```

in Python:

```python
import woothee
woothee.parse("Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 6.1; Trident/4.0)")
# => {'name': 'Internet Explorer', 'category': 'pc', 'os': 'Windows 7', 'version': '8.0', 'vendor': 'Microsoft'}
```

in Javascript(HTML, copy from release/woothee.js)
```html
<script src="./your/own/path/woothee.js"></script>
<script>
woothee.parse('Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 6.1; Trident/4.0)')
// => {name: 'Internet Explorer', category: 'pc', os: 'Windows 7', version: '8.0', vendor: 'Microsoft'}
</script>
```

in Node.js (npm install woothee)

```javascript
var woothee = require('woothee');
woothee.parse('Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 6.1; Trident/4.0)')
// => {name: 'Internet Explorer', category: 'pc', os: 'Windows 7', version: '8.0', vendor: 'Microsoft'}
```

### Hive UDF

    // add woothee.jar to classpath
    add jar woothee.jar;
    
    create temporary function parse_agent as 'is.tagomor.woothee.hive.ParseAgent';
    create temporary function is_pc as 'is.tagomor.woothee.hive.IsPC';
    create temporary function is_smartphone as 'is.tagomor.woothee.hive.IsSmartPhone';
    create temporary function is_mobilephone as 'is.tagomor.woothee.hive.IsMobilePhone';
    create temporary function is_appliance as 'is.tagomor.woothee.hive.IsAppliance';
    create temporary function is_crawler as 'is.tagomor.woothee.hive.IsCrawler';
    create temporary function is_misc as 'is.tagomor.woothee.hive.IsMisc';
    create temporary function is_unknown as 'is.tagomor.woothee.hive.IsUnknown';
    create temporary function is_in as 'is.tagomor.woothee.hive.IsIn';
    create temporary function pc as 'is.tagomor.woothee.hive.PC';
    create temporary function smartphone as 'is.tagomor.woothee.hive.SmartPhone';
    create temporary function mobilephone as 'is.tagomor.woothee.hive.MobilePhone';
    create temporary function appliance as 'is.tagomor.woothee.hive.Appliance';
    create temporary function crawler as 'is.tagomor.woothee.hive.Crawler';
    create temporary function misc as 'is.tagomor.woothee.hive.Misc';
    create temporary function unknown as 'is.tagomor.woothee.hive.Unknown';
    create temporary function oneof as 'is.tagomor.woothee.hive.OneOf';
    
    SELECT
      count(pc(parsed_agent)) as pc_pageviews,
      count(oneof(parsed_agent, array('pc', 'mobilephone', 'smartphone', 'appliance'))) as total_pageviews
    FROM (
      SELECT parse_agent(useragent) as parsed_agent FROM access_log WHERE date='today'
    ) x
    WHERE NOT is_crawler(parsed_agent) AND NOT is_unknown(parsed_agent)

## Todo

* 'mobilephone' means Japanese mobile phone groups
  * For multi-region code, domestic pattern specifier (or another mechanism) needed

## FAQ

* What's Woothee?
    * http://en.wikipedia.org/wiki/Usui_Pass
    * http://ja.wikipedia.org/wiki/%E7%A2%93%E6%B0%B7%E5%B3%A0

* * * * *

## Authors

* TAGOMORI Satoshi <tagomoris@gmail.com>
* UEDA Tetsuhiro (najeira)

## License

Copyright 2012- TAGOMORI Satoshi (tagomoris)

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


perl/Changes
Revision history for Perl module Woothee

0.3.4 Tue Sep 10 15:25:33 JST 2013
   - add Firefox OS

0.3.3  Mon Sep  2 20:12:34 JST 2013
   - add MSIE11 and Windows 8.1

0.3.1  Fri Feb  1 12:33:41 JST 2013
   - fix Playstation3 pattern
   - fix/add some crawler pattern

0.3.0  Mon Nov 26 16:13:13 JST 2012
   - add WiiU (with js implementation)

0.2.6  Wed Nov 14 11:47:17 JST 2012
   - remove perl 5.10 dependency

0.2.5  Mon Oct 22 12:41:06 JST 2012
   - update bundled Module::Install to 1.06

0.2.2  Fri Jul 13 12:51:26 JST 2012
   - Perl documents maintainance dist

0.2.1  Fri Jul  6 20:42:39 JST 2012
   - CPAN 1st dist



js/lib/dataset.js
var KEY_LABEL = exports.KEY_LABEL = 'label',
    KEY_NAME = exports.KEY_NAME = 'name',
    KEY_TYPE = exports.KEY_TYPE = 'type',
    KEY_CATEGORY = exports.KEY_CATEGORY = 'category',
    KEY_OS = exports.KEY_OS = 'os',
    KEY_VENDOR = exports.KEY_VENDOR = 'vendor',
    KEY_VERSION = exports.KEY_VERSION = 'version';

var TYPE_BROWSER = 'browser',
    TYPE_OS = 'os',
    TYPE_FULL = 'full';

var CATEGORY_PC = exports.CATEGORY_PC = 'pc',
    CATEGORY_SMARTPHONE = exports.CATEGORY_SMARTPHONE = 'smartphone',
    CATEGORY_MOBILEPHONE = exports.CATEGORY_MOBILEPHONE = 'mobilephone',
    CATEGORY_CRAWLER = exports.CATEGORY_CRAWLER = 'crawler',
    CATEGORY_APPLIANCE = exports.CATEGORY_APPLIANCE = 'appliance',
    CATEGORY_MISC = exports.CATEGORY_MISC = 'misc';

var ATTRIBUTE_NAME = exports.ATTRIBUTE_NAME = 'name',
    ATTRIBUTE_CATEGORY = exports.ATTRIBUTE_CATEGORY = 'category',
    ATTRIBUTE_OS = exports.ATTRIBUTE_OS = 'os',
    ATTRIBUTE_VENDOR = exports.ATTRIBUTE_VENDOR = 'vendor',
    ATTRIBUTE_VERSION = exports.ATTRIBUTE_VERSION = 'version';

var VALUE_UNKNOWN = exports.VALUE_UNKNOWN = 'UNKNOWN';

var CATEGORY_LIST = exports.CATEGORY_LIST = [
  CATEGORY_PC, CATEGORY_SMARTPHONE, CATEGORY_MOBILEPHONE,
  CATEGORY_CRAWLER, CATEGORY_APPLIANCE, CATEGORY_MISC, VALUE_UNKNOWN
];
var ATTRIBUTE_LIST = exports.ATTRIBUTE_LIST = [ATTRIBUTE_NAME, ATTRIBUTE_CATEGORY, ATTRIBUTE_OS, ATTRIBUTE_VENDOR, ATTRIBUTE_VERSION];

var DATASET = {};

// GENERATED from dataset.yaml at Tue Sep 10 14:45:26 JST 2013 by tagomoris
var obj;
obj = {label:'MSIE', name:'Internet Explorer', type:'browser'};
obj['vendor'] = 'Microsoft';
DATASET[obj.label] = obj;
obj = {label:'Chrome', name:'Chrome', type:'browser'};
obj['vendor'] = 'Google';
DATASET[obj.label] = obj;
obj = {label:'Safari', name:'Safari', type:'browser'};
obj['vendor'] = 'Apple';
DATASET[obj.label] = obj;
obj = {label:'Firefox', name:'Firefox', type:'browser'};
obj['vendor'] = 'Mozilla';
DATASET[obj.label] = obj;
obj = {label:'Opera', name:'Opera', type:'browser'};
obj['vendor'] = 'Opera';
DATASET[obj.label] = obj;
obj = {label:'Sleipnir', name:'Sleipnir', type:'browser'};
obj['vendor'] = 'Fenrir Inc.';
DATASET[obj.label] = obj;
obj = {label:'Win', name:'Windows UNKNOWN Ver', type:'os'};
obj['category'] = 'pc';
DATASET[obj.label] = obj;
obj = {label:'Win8.1', name:'Windows 8.1', type:'os'};
obj['category'] = 'pc';
DATASET[obj.label] = obj;
obj = {label:'Win8', name:'Windows 8', type:'os'};
obj['category'] = 'pc';
DATASET[obj.label] = obj;
obj = {label:'Win7', name:'Windows 7', type:'os'};
obj['category'] = 'pc';
DATASET[obj.label] = obj;
obj = {label:'WinVista', name:'Windows Vista', type:'os'};
obj['category'] = 'pc';
DATASET[obj.label] = obj;
obj = {label:'WinXP', name:'Windows XP', type:'os'};
obj['category'] = 'pc';
DATASET[obj.label] = obj;
obj = {label:'Win2000', name:'Windows 2000', type:'os'};
obj['category'] = 'pc';
DATASET[obj.label] = obj;
obj = {label:'WinNT4', name:'Windows NT 4.0', type:'os'};
obj['category'] = 'pc';
DATASET[obj.label] = obj;
obj = {label:'WinMe', name:'Windows Me', type:'os'};
obj['category'] = 'pc';
DATASET[obj.label] = obj;
obj = {label:'Win98', name:'Windows 98', type:'os'};
obj['category'] = 'pc';
DATASET[obj.label] = obj;
obj = {label:'Win95', name:'Windows 95', type:'os'};
obj['category'] = 'pc';
DATASET[obj.label] = obj;
obj = {label:'WinPhone', name:'Windows Phone OS', type:'os'};
obj['category'] = 'smartphone';
DATASET[obj.label] = obj;
obj = {label:'WinCE', name:'Windows CE', type:'os'};
obj['category'] = 'smartphone';
DATASET[obj.label] = obj;
obj = {label:'OSX', name:'Mac OSX', type:'os'};
obj['category'] = 'pc';
DATASET[obj.label] = obj;
obj = {label:'MacOS', name:'Mac OS Classic', type:'os'};
obj['category'] = 'pc';
DATASET[obj.label] = obj;
obj = {label:'Linux', name:'Linux', type:'os'};
obj['category'] = 'pc';
DATASET[obj.label] = obj;
obj = {label:'BSD', name:'BSD', type:'os'};
obj['category'] = 'pc';
DATASET[obj.label] = obj;
obj = {label:'Android', name:'Android', type:'os'};
obj['category'] = 'smartphone';
DATASET[obj.label] = obj;
obj = {label:'iPhone', name:'iPhone', type:'os'};
obj['category'] = 'smartphone';
DATASET[obj.label] = obj;
obj = {label:'iPad', name:'iPad', type:'os'};
obj['category'] = 'smartphone';
DATASET[obj.label] = obj;
obj = {label:'iPod', name:'iPod', type:'os'};
obj['category'] = 'smartphone';
DATASET[obj.label] = obj;
obj = {label:'iOS', name:'iOS', type:'os'};
obj['category'] = 'smartphone';
DATASET[obj.label] = obj;
obj = {label:'FirefoxOS', name:'Firefox OS', type:'os'};
obj['category'] = 'smartphone';
DATASET[obj.label] = obj;
obj = {label:'BlackBerry', name:'BlackBerry', type:'os'};
obj['category'] = 'smartphone';
DATASET[obj.label] = obj;
obj = {label:'docomo', name:'docomo', type:'full'};
obj['vendor'] = 'docomo';
obj['category'] = 'mobilephone';
obj['os'] = 'docomo';
DATASET[obj.label] = obj;
obj = {label:'au', name:'au by KDDI', type:'full'};
obj['vendor'] = 'au';
obj['category'] = 'mobilephone';
obj['os'] = 'au';
DATASET[obj.label] = obj;
obj = {label:'SoftBank', name:'SoftBank Mobile', type:'full'};
obj['vendor'] = 'SoftBank';
obj['category'] = 'mobilephone';
obj['os'] = 'SoftBank';
DATASET[obj.label] = obj;
obj = {label:'willcom', name:'WILLCOM', type:'full'};
obj['vendor'] = 'WILLCOM';
obj['category'] = 'mobilephone';
obj['os'] = 'WILLCOM';
DATASET[obj.label] = obj;
obj = {label:'jig', name:'jig browser', type:'full'};
obj['vendor'] = '';
obj['category'] = 'mobilephone';
obj['os'] = 'jig';
DATASET[obj.label] = obj;
obj = {label:'emobile', name:'emobile', type:'full'};
obj['vendor'] = '';
obj['category'] = 'mobilephone';
obj['os'] = 'emobile';
DATASET[obj.label] = obj;
obj = {label:'SymbianOS', name:'SymbianOS', type:'full'};
obj['vendor'] = '';
obj['category'] = 'mobilephone';
obj['os'] = 'SymbianOS';
DATASET[obj.label] = obj;
obj = {label:'MobileTranscoder', name:'Mobile Transcoder', type:'full'};
obj['vendor'] = '';
obj['category'] = 'mobilephone';
obj['os'] = 'Mobile Transcoder';
DATASET[obj.label] = obj;
obj = {label:'Nintendo3DS', name:'Nintendo 3DS', type:'full'};
obj['vendor'] = 'Nintendo';
obj['category'] = 'appliance';
obj['os'] = 'Nintendo 3DS';
DATASET[obj.label] = obj;
obj = {label:'NintendoDSi', name:'Nintendo DSi', type:'full'};
obj['vendor'] = 'Nintendo';
obj['category'] = 'appliance';
obj['os'] = 'Nintendo DSi';
DATASET[obj.label] = obj;
obj = {label:'NintendoWii', name:'Nintendo Wii', type:'full'};
obj['vendor'] = 'Nintendo';
obj['category'] = 'appliance';
obj['os'] = 'Nintendo Wii';
DATASET[obj.label] = obj;
obj = {label:'NintendoWiiU', name:'Nintendo Wii U', type:'full'};
obj['vendor'] = 'Nintendo';
obj['category'] = 'appliance';
obj['os'] = 'Nintendo Wii U';
DATASET[obj.label] = obj;
obj = {label:'PSP', name:'PlayStation Portable', type:'full'};
obj['vendor'] = 'Sony';
obj['category'] = 'appliance';
obj['os'] = 'PlayStation Portable';
DATASET[obj.label] = obj;
obj = {label:'PSVita', name:'PlayStation Vita', type:'full'};
obj['vendor'] = 'Sony';
obj['category'] = 'appliance';
obj['os'] = 'PlayStation Vita';
DATASET[obj.label] = obj;
obj = {label:'PS3', name:'PlayStation 3', type:'full'};
obj['vendor'] = 'Sony';
obj['category'] = 'appliance';
obj['os'] = 'PlayStation 3';
DATASET[obj.label] = obj;
obj = {label:'DigitalTV', name:'InternetTVBrowser', type:'full'};
obj['vendor'] = '';
obj['category'] = 'appliance';
obj['os'] = 'DigitalTV';
DATASET[obj.label] = obj;
obj = {label:'SafariRSSReader', name:'Safari RSSReader', type:'full'};
obj['vendor'] = 'Apple';
obj['category'] = 'misc';
DATASET[obj.label] = obj;
obj = {label:'GoogleDesktop', name:'Google Desktop', type:'full'};
obj['vendor'] = 'Google';
obj['category'] = 'misc';
DATASET[obj.label] = obj;
obj = {label:'WindowsRSSReader', name:'Windows RSSReader', type:'full'};
obj['vendor'] = 'Microsoft';
obj['category'] = 'misc';
DATASET[obj.label] = obj;
obj = {label:'VariousRSSReader', name:'RSSReader', type:'full'};
obj['vendor'] = '';
obj['category'] = 'misc';
DATASET[obj.label] = obj;
obj = {label:'HTTPLibrary', name:'HTTP Library', type:'full'};
obj['vendor'] = '';
obj['category'] = 'misc';
DATASET[obj.label] = obj;
obj = {label:'GoogleBot', name:'Googlebot', type:'full'};
obj['vendor'] = '';
obj['category'] = 'crawler';
DATASET[obj.label] = obj;
obj = {label:'GoogleBotMobile', name:'Googlebot Mobile', type:'full'};
obj['vendor'] = '';
obj['category'] = 'crawler';
DATASET[obj.label] = obj;
obj = {label:'GoogleMediaPartners', name:'Google Mediapartners', type:'full'};
obj['vendor'] = '';
obj['category'] = 'crawler';
DATASET[obj.label] = obj;
obj = {label:'GoogleFeedFetcher', name:'Google Feedfetcher', type:'full'};
obj['vendor'] = '';
obj['category'] = 'crawler';
DATASET[obj.label] = obj;
obj = {label:'GoogleAppEngine', name:'Google AppEngine', type:'full'};
obj['vendor'] = '';
obj['category'] = 'crawler';
DATASET[obj.label] = obj;
obj = {label:'GoogleWebPreview', name:'Google Web Preview', type:'full'};
obj['vendor'] = '';
obj['category'] = 'crawler';
DATASET[obj.label] = obj;
obj = {label:'YahooSlurp', name:'Yahoo! Slurp', type:'full'};
obj['vendor'] = '';
obj['category'] = 'crawler';
DATASET[obj.label] = obj;
obj = {label:'YahooJP', name:'Yahoo! Japan', type:'full'};
obj['vendor'] = '';
obj['category'] = 'crawler';
DATASET[obj.label] = obj;
obj = {label:'YahooPipes', name:'Yahoo! Pipes', type:'full'};
obj['vendor'] = '';
obj['category'] = 'crawler';
DATASET[obj.label] = obj;
obj = {label:'Baiduspider', name:'Baiduspider', type:'full'};
obj['vendor'] = '';
obj['category'] = 'crawler';
DATASET[obj.label] = obj;
obj = {label:'msnbot', name:'msnbot', type:'full'};
obj['vendor'] = '';
obj['category'] = 'crawler';
DATASET[obj.label] = obj;
obj = {label:'bingbot', name:'bingbot', type:'full'};
obj['vendor'] = '';
obj['category'] = 'crawler';
DATASET[obj.label] = obj;
obj = {label:'Yeti', name:'Naver Yeti', type:'full'};
obj['vendor'] = '';
obj['category'] = 'crawler';
DATASET[obj.label] = obj;
obj = {label:'FeedBurner', name:'Google FeedBurner', type:'full'};
obj['vendor'] = '';
obj['category'] = 'crawler';
DATASET[obj.label] = obj;
obj = {label:'facebook', name:'facebook', type:'full'};
obj['vendor'] = '';
obj['category'] = 'crawler';
DATASET[obj.label] = obj;
obj = {label:'mixi', name:'mixi', type:'full'};
obj['vendor'] = '';
obj['category'] = 'crawler';
DATASET[obj.label] = obj;
obj = {label:'IndyLibrary', name:'Indy Library', type:'full'};
obj['vendor'] = '';
obj['category'] = 'crawler';
DATASET[obj.label] = obj;
obj = {label:'ApplePubSub', name:'Apple iCloud', type:'full'};
obj['vendor'] = '';
obj['category'] = 'crawler';
DATASET[obj.label] = obj;
obj = {label:'Genieo', name:'Genieo Web Filter', type:'full'};
obj['vendor'] = '';
obj['category'] = 'crawler';
DATASET[obj.label] = obj;
obj = {label:'topsyButterfly', name:'topsy Butterfly', type:'full'};
obj['vendor'] = '';
obj['category'] = 'crawler';
DATASET[obj.label] = obj;
obj = {label:'rogerbot', name:'SeoMoz rogerbot', type:'full'};
obj['vendor'] = '';
obj['category'] = 'crawler';
DATASET[obj.label] = obj;
obj = {label:'AhrefsBot', name:'ahref AhrefsBot', type:'full'};
obj['vendor'] = '';
obj['category'] = 'crawler';
DATASET[obj.label] = obj;
obj = {label:'radian6', name:'salesforce radian6', type:'full'};
obj['vendor'] = '';
obj['category'] = 'crawler';
DATASET[obj.label] = obj;
obj = {label:'Hatena', name:'Hatena', type:'full'};
obj['vendor'] = '';
obj['category'] = 'crawler';
DATASET[obj.label] = obj;
obj = {label:'goo', name:'goo', type:'full'};
obj['vendor'] = '';
obj['category'] = 'crawler';
DATASET[obj.label] = obj;
obj = {label:'livedoorFeedFetcher', name:'livedoor FeedFetcher', type:'full'};
obj['vendor'] = '';
obj['category'] = 'crawler';
DATASET[obj.label] = obj;
obj = {label:'VariousCrawler', name:'misc crawler', type:'full'};
obj['vendor'] = '';
obj['category'] = 'crawler';
DATASET[obj.label] = obj;

var get = exports.get = function(label) {
  return DATASET[label];
};
